helix-lsp/src/lib.rs

```diff
@@ -605,109 +599,118 @@ impl Notification {
+                if self.inner.contains_key(&name) {
+                    let client = match self.start_client(name.clone(), &lang_config.roots, doc_path)
+                    {
+                        Ok(client) => client,
+                        error => return Some(error),
+                    };
 
-                let (_, old_client) = entry.insert((id, client.clone()));
+                    let old_client = self.inner.insert(name, client.clone()).unwrap();
 
-                tokio::spawn(async move {
-                    let _ = old_client.force_shutdown().await;
-                });
+                    tokio::spawn(async move {
+                        let _ = old_client.force_shutdown().await;
+                    });
 
-                Ok(Some(client))
-            }
-        }
+                    Some(Ok(client))
+                } else {
+                    None
+                }
+            })
+            .collect()
     }
 
-    pub fn stop(&mut self, language_config: &LanguageConfiguration) {
-        let scope = language_config.scope.clone();
-
-        if let Some((_, client)) = self.inner.remove(&scope) {
-            tokio::spawn(async move {
-                let _ = client.force_shutdown().await;
-            });
+    pub fn stop(&mut self, name: &str) {
+        if let Some(client) = self.inner.remove(name) {
+            tokio::spawn(async move { client.force_shutdown().await.ok() });
         }
     }
 
     pub fn get(
         &mut self,
-        language_config: &LanguageConfiguration,
+        lang_config: &LanguageConfiguration,
         doc_path: Option<&std::path::PathBuf>,
-    ) -> Result<Option<Arc<Client>>> {
-        let config = match &language_config.language_server {
-            Some(config) => config,
-            None => return Ok(None),
-        };
-
-        match self.inner.entry(language_config.scope.clone()) {
-            Entry::Occupied(entry) => Ok(Some(entry.get().1.clone())),
-            Entry::Vacant(entry) => {
-                // initialize a new client
-                let id = self.counter.fetch_add(1, Ordering::Relaxed);
-
-                let NewClientResult(client, incoming) =
-                    start_client(id, language_config, config, doc_path)?;
-                self.incoming.push(UnboundedReceiverStream::new(incoming));
-
-                entry.insert((id, client.clone()));
-                Ok(Some(client))
-            }
-        }
+    ) -> Result<Vec<Arc<Client>>> {
+        lang_config
+            .language_servers
+            .iter()
+            .map(|config| match self.inner.get(config.name()) {
+                Some(client) => Ok(client.clone()),
+                None => {
+                    let client =
+                        self.start_client(config.name().clone(), &lang_config.roots, doc_path)?;
+                    self.inner.insert(config.name().clone(), client.clone());
+                    Ok(client)
+                }
+            })
```

Comment:
Same, why not use HashMap entry API here?

Reply:
See above.